HARK/ConsumptionSaving/ConsRiskyContribModel.py

```diff
@@ -0,0 +1,1768 @@
+    """
+    A consumer type with idiosyncratic shocks to permanent and transitory income,
+    who can save in both a risk-free and a risky asset but faces frictions to
+    moving funds between them. The agent can only consume out of his risk-free
+    asset.
+    
+    The frictions are:
+        - A proportional tax on funds moved from the risky to the risk-free
+         asset.
+        - A stochastic inability to move funds between his accounts.
+    
+    To partially avoid the second friction, the agent can commit to have a
+    fraction of his labor income, which is usually deposited in his risk-free
+    account, diverted to his risky account. He can change this fraction
+    only in periods where he is able to move funds between accounts.
+    """
+
+    time_inv_ = deepcopy(RiskyAssetConsumerType.time_inv_)
+    time_inv_ = time_inv_ + ["DiscreteShareBool"]
+
+    state_vars = RiskyAssetConsumerType.state_vars + [
+        "nNrm",
+        "mNrmTilde",
+        "nNrmTilde",
+        "Share",
+    ]
+    shock_vars_ = RiskyAssetConsumerType.shock_vars_
+
+    def __init__(self, cycles=1, verbose=False, quiet=False, **kwds):
+
+        params = init_risky_contrib.copy()
+        params.update(kwds)
+        kwds = params
+
+        # Initialize a basic consumer type
+        RiskyAssetConsumerType.__init__(
+            self, cycles=cycles, verbose=verbose, quiet=quiet, **kwds
+        )
+
+        # The model is solved and simulated spliting each of the agent's
+        # decisions into its own "stage". The stages in chronological order
+        # are
+        # - Reb: asset-rebalancing stage.
+        # - Sha: definition of the income contribution share.
+        # - Cns: consumption stage.
+        self.stages = ["Reb", "Sha", "Cns"]
+
+        # Each stage has its own states and controls, and its methods
+        # to find them.
+        self.set_states = {
+            "Reb": self.get_states_Reb,
+            "Sha": self.get_states_Sha,
+            "Cns": self.get_states_Cns,
+        }
+
+        self.set_controls = {
+            "Reb": self.get_controls_Reb,
+            "Sha": self.get_controls_Sha,
+            "Cns": self.get_controls_Cns,
+        }
+
+        # Set the solver for the portfolio model, and update various constructed attributes
+        self.solve_one_period = solveRiskyContrib
+        self.update()
+
+    def pre_solve(self):
+        self.update_solution_terminal()
+
+    def update(self):
+
+        RiskyAssetConsumerType.update(self)
+        self.update_share_grid()
+        self.update_dfrac_grid()
+        self.update_nNrm_grid()
+        self.update_mNrm_grid()
+        self.update_tau()
+
+    def update_solution_terminal(self):
+        """
+        Solves the terminal period. The solution is trivial.
+        Cns: agent will consume all of his liquid resources.
+        Sha: irrelevant as there is no "next" period.
+        Reb: agent will shift all of his resources to the risk-free asset.
```

Comment:
It's a little inelegant to have the stage functionality separated into different methods for states and controls, but not separated out in this method of computing the solution terminals.

Reply:
I agree, but the trivial nature of the terminal solution makes it much easier to construct it all at once, because some objects can be re-used.
Also this trivial stage-solutions are created only once throughout the code, and must be created together. Therefore I don't see much gain from defining `update_solution_terminal_(Reb|Sha|Cns)()`.
I can do it if you feel strongly about it.